stdlib/bytes.gr

```diff
@@ -0,0 +1,466 @@
+ * If the `left` or `right` parameters are negative, bytes will be removed from the corresponding side instead of appended.
+ * @param left: Number - The number of uninitialized bytes to prepend
+ * @param right: Number - The number of uninitialized bytes to append
+ * @param b: Bytes - The Bytes buffer get a subset of bytes from
+ * @returns Bytes
+ */
+@disableGC
+export let extend = (left: Number, right: Number, b: Bytes) => {
+  let (<) = WasmI32.ltS
+  let (>) = WasmI32.gtS
+  let (+) = WasmI32.add
+  let (-) = WasmI32.sub
+  let (*) = WasmI32.mul
+  let src = WasmI32.fromGrain(b)
+  let size = getSize(src)
+  let left = numberToWasmI32(left)
+  let right = numberToWasmI32(right)
+  let newSize = size + left + right
+  if (newSize < 0n) {
+    throw Exception.InvalidArgument("The resulting length is less than 0")
+  }
+  let dst = allocateBytes(newSize)
+  let mut srcOffset = 0n;
+  let mut dstOffset = 0n;
+  if (left < 0n) {
+    srcOffset = left * -1n;
+    dstOffset = 0n;
+  }
+  if (left > 0n) {
+    srcOffset = 0n;
+    dstOffset = left;
+  }
+  let len = if (right < 0n) {
+    (size + right) - srcOffset
+  } else {
+    size - srcOffset
+  }
+  if (len > 0n) {
+    Memory.copy(dst + _VALUE_OFFSET + dstOffset, src + _VALUE_OFFSET + srcOffset, len)
+  }
+  WasmI32.toGrain(dst): Bytes
+}
+
+/**
+ * Copies a range of bytes bytes from a source buffer to a given location in a destination buffer.
+ * @param srcPos: Number - The starting byte index to copy bytes from
+ * @param dstPos: Number - The starting byte index to copy bytes into
+ * @param len: Number - The amount of bytes to copy from the source buffer
+ * @param src: Bytes - The source buffer
+ * @param dst: Bytes - The destination buffer
+ * @returns Void
+ */
+@disableGC
+export let move = (srcPos: Number, dstPos: Number, len: Number, src: Bytes, dst: Bytes) => {
+  let (>) = WasmI32.gtS
+  let (+) = WasmI32.add
+  let src = WasmI32.fromGrain(src)
+  let srcSize = getSize(src)
+  let srcPos = numberToWasmI32(srcPos)
+  let dst = WasmI32.fromGrain(dst)
+  let dstSize = getSize(dst)
+  let dstPos = numberToWasmI32(dstPos)
+  let len = numberToWasmI32(len)
+  if ((srcPos + len) > srcSize) {
+    throw Exception.InvalidArgument("Invalid source bytes range")
+  }
+  if ((dstPos + len) > dstSize) {
+    throw Exception.InvalidArgument("Invalid destination bytes range")
+  }
+  let end = srcPos + len
+  Memory.copy(dst + _VALUE_OFFSET + dstPos, src + _VALUE_OFFSET + srcPos, len)
+}
+
+/**
+ * Get the byte length of a Bytes buffer.
+ * @param b: Bytes - The Bytes buffer to check
+ * @returns Number
+ */
+ @disableGC
+export let length = (b: Bytes) => {
+  let b = WasmI32.fromGrain(b)
+  tagSimpleNumber(getSize(b))
+}
+
+/**
+ * Creates a new Bytes buffer that contains the bytes of both buffers a and b.
+ * @param a: Bytes - The buffer to be copied first
+ * @param b: Bytes - The buffer to be copied last
+ * @returns Bytes
+ */
+export let append = (a: Bytes, b: Bytes) => {
```

Comment:
No `prepend`?

Reply:
That's just the same but with the arguments reversed, no?

Reply:
Sure, but then the name just seems weird. Shall we use `concat` instead?